### ISSUE TITLE
Problems in `ArchLinux` with `vscode`

### ISSUE THREAD

**Issue by sarpik** [jorangreef/sudo-prompt#113]
Hello,

This package is used in [vscode](https://github.com/microsoft/vscode).
In [arch linux](https://www.archlinux.org/), we're unable to use the `Save as sudo` functionality though `vscode`.

Please see https://github.com/microsoft/vscode/issues/82266 -- it's the original issue, providing other related issues, specific details & possible candidates for a fix.

Thank you!

**Comment by jorangreef** [comment]
Thanks @sarpik, ArchLinux has been problematic compared to other distributions.

I have looked through https://github.com/microsoft/vscode/issues/82266, and need to follow all those links. If you have time, would you mind summarizing here?

If not, it's fine, I will get to this, I just need some time.

**Comment by sarpik** [comment]
No problem @jorangreef,

I've tried summarizing, but pretty much everything is documented in https://github.com/microsoft/vscode/issues/82266. For a quick summary, I believe you can read through it without following the links - we tried providing context from them.

<details>

<summary>
The stuff I wrote that is pretty much the same, you'd be better off just going through https://github.com/microsoft/vscode/issues/82266 :D
</summary>


I'll try my best to summarize (I'll provide links just in case):

It seems like the issue was first reported here: https://bugs.archlinux.org/task/61516

> When trying to save a file with insufficient permissions, vscode pops up a notification.
You can now "retry with sudo".
However, this does not work.

And then later in https://github.com/microsoft/vscode/issues/70403 and I see a few comments from you there.

The conclusion was reached that it's an upstream [arch linux] issue, and that installing the Microsoft's packaged binary version of vscode solves the issue: https://aur.archlinux.org/packages/visual-studio-code-bin/

---

Then I found all of this, because I had the same issue, and wanted it fixed.
I noticed that the vscode team blames upstream [arch linux], but arch linux then blames upstream [vscode].

This wouldn't get the issue fixed, so I created an issue to bring attention: https://github.com/microsoft/vscode/issues/82266. There's useful information, and it's the latest issue related to this.

They recommended me to report this issue here, since this package is used for vscode's save as sudo feature, which is what we're having issues with.

</details>

**Comment by jorangreef** [comment]
Thanks for your help with all of this @sarpik, I would also like to get this fixed, if it's something we can fix.

What I wanted to know in particular, was why installing the packaged binary version of VS Code solved the issue? If it did, that makes it seem like it's not VS Code or sudo-prompt, unless the packaged binary version of VS Code was a newer version with a different version of sudo-prompt. If you could confirm that the versions of VS Code were exactly the same between packaged binary and non-packaged that would help me alot!

The comment by ArchLinux does sound like a recipe for disaster:

> We [arch linux] are sane people so we don't distribute a full copy of electron with VS Code, we use the electron already present on your system. 

I will dig into everything and see what we can do!

**Comment by sarpik** [comment]
Thanks Joran!

I think the problem might be with the packaging of electron, as seen in [the comment from arch linux](https://bugs.archlinux.org/task/61516#comment178392) which you just quoted, but I'm just speculating.

I will try to install the packaged binary right now, and will update you soon.

Sorry to bother @FFY00, but I'm pretty sure you know the most out of us here. Any ideas?

**Comment by FFY00** [comment]
Basically, we start vscode like this:

```bash
ELECTRON_RUN_AS_NODE=1 exec electron4 /usr/lib/code/out/cli.js /usr/lib/code/code.js "$@"
```

**Comment by sarpik** [comment]
Alright, I've tried out 3 things:

1. code (oss, not binary)
2. vscodium (oss, binary)
3. code (not oss (microsoft's release), binary)

(all versions identical)

TL;DR:

`Save as sudo` worked with both binaries. Non-binary version did not.

---

Fyi:

The binary for arch linux is provided here: https://aur.archlinux.org/packages/visual-studio-code-bin/

and [from the FAQ comment](https://aur.archlinux.org/packages/visual-studio-code-bin/?O=0&PP=10#pinned-620944):

> What is the difference between this package and the one in the community repo?

> This is the official binary distribution from Microsoft. The one in the community repo is an unofficial build made from source. Beyond the license difference and branding, there are some proprietary features not available in the open source version.

---

1. I'm trying this out with the [oss vscode from arch linux (not the binary)](https://www.archlinux.org/packages/?name=code).

(see https://wiki.archlinux.org/index.php/Visual_Studio_Code#Installation for the different versions)

Version:

```sh
# code
$ code --version
1.39.2
6ab598523be7a800d7f3eb4d92d7ab9a66069390
x64
```

First attempt at editing a file without write permission:

```
Failed to save 'xorg.conf': No polkit authentication agent found.
```

I stumbled upon this https://github.com/balena-io/etcher/issues/1179#issuecomment-285912149 and then, from the [arch wiki page about Polkit](https://wiki.archlinux.org/index.php/Polkit), I installed [mate-polkit](https://www.archlinux.org/packages/community/x86_64/mate-polkit/).

I needed to run it:

```sh
$ /usr/lib/mate-polkit/polkit-mate-authentication-agent-1
```

and then tried to save the file without write permissions again:

[I got the prompt for my password](https://i.imgur.com/FG1fdqv.png), but after authenticating successfully, I got this from vscode's pop-up:

```console
Failed to save 'xorg.conf': Command failed: cd "/home/kipras"; "/usr/bin/pkexec" --disable-internal-agent /bin/bash -c "echo SUDOPROMPT; \"/usr/bin/code\" --file-write \"/tmp/code-elevated-rawbag\" \"/etc/X11/xorg.conf\"" Error using --file-write: EACCES: permission denied, open '/tmp/code-elevated-rawbag'
```

---

2. I've just tried the same thing with [vscodium](https://vscodium.com/) (in arch, [vscodium-bin from AUR](https://aur.archlinux.org/packages/vscodium-bin/)) (I uninstalled `code` beforehand).

It IS a binary release, and everything WORKED!

```sh
# vscodium-bin
$ codium --version
1.39.2
6ab598523be7a800d7f3eb4d92d7ab9a66069390
x64
```

---

3. Just tried out the same thing with [microsoft's vscode binary](https://aur.archlinux.org/packages/visual-studio-code-bin/)

The results are the same as with the vscodium binary - everything worked fine.

```sh
# visual-studio-code-bin
$ code --version
1.39.2
6ab598523be7a800d7f3eb4d92d7ab9a66069390
x64
```

---

Sorry to bother you too, @bpasero, but maybe you have any ideas about this?

**Comment by jorangreef** [comment]
> Holy moly, I think I'm sticking with vscodium-bin :D

😄 Thanks @sarpik, that's a fantastic write-up.

Do you still think this is a sudo-prompt or VS Code issue then?

**Comment by jorangreef** [comment]
Maybe there's something we could do though to help ArchLinux out? Any ideas?

**Comment by sarpik** [comment]
> > Holy moly, I think I'm sticking with vscodium-bin :D
> 
>  Thanks @sarpik, that's a fantastic write-up.

lmao, thank you for being awesome 😄

> Do you still think this is a sudo-prompt or VS Code issue then?

yeah I don't know, the only difference I see is whether vscode is packaged in binary form or not. This is probably to do with arch linux, but I'm not qualified to answer:D

I'd love to help arch linux out, but honestly, I have no clue how.
Any last thoughts, @FFY00? c:

**Comment by nova-nowiz** [comment]
I have the same problem on NixOS, I can't save files as sudo in vscode.
However, I can use `pkexec code --file-write /path/to/source /path/to/target` and it works fine, I'm confused...
Should I report this on vscode issue tracker?
